### PR TITLE
fix: more efficient sender with gossiped decisions

### PR DESCRIPTION
**Pull request by kentquirk**
## Which problem is this PR solving?

- The current sender is inefficient; it keeps sending a large list of trace IDs to Redis to ask for decisions; most of them aren't current. The result is a lot of traffic with Redis, most of which is useless.

## Short description of the changes

- Create two new gossip channels (keep and send)
- When a decision is made, gossip the decision
- The gossip listeners accumulate batches of keep or drop decisions
    - drop is dropped immediately with no other traffic
    - keep gets statuses and acts the same way the existing sender does
- The accumulator uses an aggregation algorithm that collects messages for a max of time or number of items, then hands them off to the sender for dispatch in a goroutine.
- If the publish outraced the propagation of the trace status, it tries again next time.
- Renamed GetOldest to GetHighImpact to clarify what it does.
- Implement SpanCache.GetOldest (with test) that returns the items in the cache that have been hanging around longer than 2*(TraceTimeout+SendDelay). These traces that should have been decided by now but no gossip message was received.
- There's now a cleanup task that runs once every TraceTimeout and explicitly asks only about "Oldest" traces. This might happen after a refinery gets added to a cluster, or when the decider loop is very busy. It also acts as a hedge against issues in pub/sub, which doesn't have very robust delivery guarantees.
- Add a (temporary) config flag to control the use of this feature.

The hope is that this ends up with much less redis traffic, and less memory usage overall since decided traces are dispatched more quickly (soon after decision was made). It should also reduce latency for kept traces.

Since this task required modifying the SpanCache, it also deletes the complex span cache, which was an experiment that failed. It is both slower and more complicated than the basic span cache.

It passes existing tests because we haven't deleted the sender stuff yet, but it needs tests in CentralCollector that depend on it. 

In local testing, it appears to be noticeably faster and more consistent than the sender implementation.

The hope is to test this, and if it does what's desired, to delete the existing sender implementation and its config items. This will also have an effect on the health check and stress relief.

### CODE REVIEW

File: internal/gossip/gossip.go
Comment (TylerHelmuth) [inline internal/gossip/gossip.go:46]: Logging when we do this would be helpful

File: collect/central_collector.go
Comment (TylerHelmuth) [inline collect/central_collector.go:873]: Since we still set the status below, is there a possible race condition between the status in redis and the new `keepTraces`?

Comment (kentquirk) [reply collect/central_collector.go:873]: Yes, which is why KeepTraces returns the list of things that didn't have a status available from Redis yet, so that it can be retried quickly.

Without it, it still works, but it has to wait for the cleanup task to find it, which might take several seconds. 

Noticing that delay is, in fact, why I added the retry logic.

File: internal/gossip/gossip.go
Comment (kentquirk) [reply internal/gossip/gossip.go:46]: I will, but prefer to wait until #1159 is merged before I touch gossip code. This was me just wanting to note that this could drop data.

File: collect/central_collector.go
Comment (TylerHelmuth) [inline collect/central_collector.go:641]: Does this cleanup process have an effect on the spans processed metrics? Should we be incrementing that metric since we are "reprocessing" the span?

**Review by TylerHelmuth — DISMISSED** [review]

File: collect/central_collector.go
Comment (kentquirk) [reply collect/central_collector.go:612]: See if you think this is a good solution?

**Review by TylerHelmuth — APPROVED** [review]

**Pull request merged**
